File: src/main/java/com/htmake/reader/utils/SpringContextUtils.java
package com.htmake.reader.utils;

import org.springframework.beans.BeansException;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.stereotype.Component;

@Component
public class SpringContextUtils implements ApplicationContextAware {
    private static ApplicationContext applicationContext;

    @Override
    public void setApplicationContext(ApplicationContext context) throws BeansException {
        applicationContext = context;
    }

    public static ApplicationContext getApplicationContext() {
        return applicationContext;
    }

    public static Object getBean(String name) {
        if (applicationContext != null) {
            return getApplicationContext().getBean(name);
        }
        return null;
    }

    public static <T> T getBean(Class<T> clazz) {
        if (applicationContext != null) {
            return getApplicationContext().getBean(clazz);
        }
        return null;
    }

    public static <T> T getBean(String name, Class<T> clazz) {
        if (applicationContext != null) {
            return getApplicationContext().getBean(name, clazz);
        }
        return null;
    }
}

File: src/main/kotlin/com/htmake/reader/utils/SpringContextUtils.kt
package com.htmake.reader.utils

import org.springframework.context.ApplicationContext
import org.springframework.context.ApplicationContextAware
import org.springframework.stereotype.Component

@Component
class SpringContextUtils : ApplicationContextAware {
    override fun setApplicationContext(applicationContext: ApplicationContext) {
        ctx = applicationContext
    }

    companion object {
        @Volatile private var ctx: ApplicationContext? = null

        fun <T> getBean(name: String, type: Class<T>): T =
            ctx?.getBean(name, type)
                ?: error("Spring context not ready for bean $name")

        fun <T> getBean(type: Class<T>): T =
            ctx?.getBean(type)
                ?: error("Spring context not ready for type ${type.name}")
    }
}
